packages/server-wallet/src/wallet/store.ts

```diff
@@ -1,66 +1,225 @@
-        throw Error('Not in channel');
-      }
-
-      const {address: signingAddress} = signingWallet;
-      const cols: RequiredColumns = {...signedState, vars: [addHash(signedState)], signingAddress};
-
-      channel = Channel.fromJson(cols);
-      await Channel.query().insert(channel);
-    } else {
-      signedState.signatures?.map(sig => channel.addState(signedState, sig));
-      await Channel.query().update(channel);
-    }
+    const singingWalletResult = await getSigningWallet(signedState);
+    if (isLeft(singingWalletResult)) return singingWalletResult;
+
+    const {address: signingAddress} = singingWalletResult.right;
+
+    const cols: RequiredColumns = {...signedState, vars: [addHash(signedState)], signingAddress};
+    const channel = Channel.fromJson(cols);
+
+    let channelVars = channel.vars;
+
+    const addResult = addState(channelVars, signedState);
+    if (isLeft(addResult)) throw addResult.left;
+
+    channelVars = clearOldStates(channelVars, channel.isSupported ? channel.support : undefined);
+
+    const invariantValidationResult = validateInvariants(channelVars, channel.myAddress);
+    if (isLeft(invariantValidationResult)) return invariantValidationResult;
+
+    await Channel.query().upsertGraph(channel);
+
+    return right(undefined);
   },
 };
+
+class StoreError extends Error {
+  readonly type = 'InvariantError';
+  constructor(reason: StoreErrors, public readonly data: any = undefined) {
+    super(reason);
+  }
+}
+
+enum StoreErrors {
+  duplicateTurnNums = 'multiple states with same turn number',
+  notSorted = 'states not sorted',
+  multipleSignedStates = 'Store signed multiple states for a single turn',
+  invalidSignature = 'Invalid signature',
+  notInChannel = 'Not in channel',
+  staleState = 'Stale state',
+}
+
+function validateSignatures(signedState: SignedState): Either<StoreError, undefined> {
+  const {participants} = signedState;
+
+  for (const sig of signedState.signatures) {
+    const signerIndex = participants.findIndex(p => p.signingAddress === sig.signer);
+    if (signerIndex === -1) {
+      return left(new StoreError(StoreErrors.invalidSignature, {signedState, signature: sig}));
+    }
+  }
+  return right(undefined);
+}
+
+async function getSigningWallet(
+  signedState: SignedState
+): Promise<Either<StoreError, SigningWallet>> {
+  const addresses = signedState.participants.map(p => p.signingAddress);
+  const signingWallet = await SigningWallet.query()
+    .whereIn('address', addresses)
+    .first();
+
+  if (!signingWallet) {
+    return left(new StoreError(StoreErrors.notInChannel));
+  }
+  return right(signingWallet);
+}
+
+function validateStateFreshness(
+  signedState: State,
+  channel: Channel
+): Either<StoreError, undefined> {
+  if (
+    channel.isSupportedByMe &&
+    channel.latestSignedByMe &&
+    channel.latestSignedByMe.turnNum >= signedState.turnNum
+  ) {
+    return left(new StoreError(StoreErrors.staleState));
+  }
+  return right(undefined);
```

Comment:
Seems more intuitive to return `true`, maybe?

Reply:
Using a `boolean` value seems slightly strange with `Either` since you would never have a `right` value of `false`. Using `undefined` seems to fit better with the pattern of calling the `validate` method and doing something only if there's an error.

Although using undefined is still slightly weird...